### PR TITLE
manual: update docs when NIX_PATH is empty

### PR DESCRIPTION
**Pull request by zuzuleinen**
Update documentation how Nix resolves <nixpkgs> when `NIX_PATH` is not set

### CODE REVIEW

File: doc/manual/src/command-ref/env-common.md [hunk 1]
```diff
@@ -44,6 +44,12 @@ Most Nix commands interpret the following environment variables:
     The Nix search path can also be extended using the `-I` option to
     many Nix commands, which takes precedence over `NIX_PATH`.
 
+    If `NIX_PATH` is not set or empty, Nix will look up the following locations for a revision of `nixpkgs`:
+
+    1. `$HOME/.nix-defexpr/channels`
+    2. `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
+    3. `/nix/var/nix/profiles/per-user/root/channels.`
```
Comment (edolstra) [inline doc/manual/src/command-ref/env-common.md:51]:
```suggestion
    3. `/nix/var/nix/profiles/per-user/root/channels`
```

File: doc/manual/src/command-ref/env-common.md [hunk 2]
```diff
@@ -44,6 +44,12 @@ Most Nix commands interpret the following environment variables:
     The Nix search path can also be extended using the `-I` option to
     many Nix commands, which takes precedence over `NIX_PATH`.
 
+    If `NIX_PATH` is not set or empty, Nix will look up the following locations for a revision of `nixpkgs`:
```
Comment (edolstra) [inline doc/manual/src/command-ref/env-common.md:47]: This is not correct. It doesn't look for `nixpkgs` - it adds these directories to the search path, regardless of whether they contain `nixpkgs`. In particular, the first and third entries can contain non-nixpkgs channels.

Comment (jtojnar) [inline doc/manual/src/command-ref/env-common.md:47]: The fallback also does not seem to be used with empty `NIX_PATH`:

```ShellSession
$ nix-channel --add https://nixos.org/channels/nixos-unstable nixpkgs
$ nix-channel --update
$ unset NIX_PATH
$ nix-instantiate -E '<nixpkgs>'
error: expression does not evaluate to a derivation (or a set or list of those)
$ NIX_PATH= nix-instantiate -E '<nixpkgs>'
error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I)
```

Comment (fricklerhandwerk) [reply doc/manual/src/command-ref/env-common.md:47]:
```suggestion
    If `NIX_PATH` is not set or empty, Nix will look up the following locations to set the search path:
```

We still have no explanations of the two mechanisms (`name=path` and how contents of a directory on the search path are mapped to `<name>` expressions), but that is out of scope here.
Eventually terms like "search path" should link to their definition, too.

Comment (zuzuleinen) [reply doc/manual/src/command-ref/env-common.md:47]: I'm not too familiar with Nix. But locally, where I have an empty NIX_PATH, this command builds the go package for me:

`nix-build --no-out-link -E 'with import <nixpkgs> { }; go'`

Comment (jtojnar) [reply doc/manual/src/command-ref/env-common.md:47]: Are you sure you have the variable set to an empty string rather than unset?

Comment (jtojnar) [reply doc/manual/src/command-ref/env-common.md:47]: Maybe something like:

> If `NIX_PATH` environment variable is not set, Nix will fall back to the following list in impure and unrestricted evaluation mode:
> 
>     1. `$HOME/.nix-defexpr/channels`
>     2. `nixpkgs=/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
>     3. `/nix/var/nix/profiles/per-user/root/channels`

Comment (zuzuleinen) [reply doc/manual/src/command-ref/env-common.md:47]: I tried:

unset NIX_PATH && nix-build --no-out-link -E 'with import <nixpkgs> { }; go' 
env NIX_PATH="" && nix-build --no-out-link -E 'with import <nixpkgs> { }; go'

and both are building to `/nix/store/n449plwz10c5l53qhfnrywyns0m8g0qv-go-1.17.11`

Comment (fricklerhandwerk) [reply doc/manual/src/command-ref/env-common.md]: @zuzuleinen You have to do
```console
 env NIX_PATH="" nix-build --no-out-link -E 'with import <nixpkgs> { }; go'
 ```
 instead. Note there is no `&&`, as just running `env` prints the environment, but running `env <assignments> <command>` runs the command with the assignments on top of the regular environment.

File: doc/manual/src/command-ref/env-common.md [hunk 3]
```diff
@@ -44,7 +44,7 @@ Most Nix commands interpret the following environment variables:
     The Nix search path can also be extended using the `-I` option to
     many Nix commands, which takes precedence over `NIX_PATH`.
 
-    If `NIX_PATH` is not set or empty, Nix will look up the following locations for a revision of `nixpkgs`:
+    If `NIX_PATH` environment variable is not set, Nix will fall back to the following list in impure and unrestricted evaluation mode:
```
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/env-common.md:47]:
```suggestion
    If `NIX_PATH` is not set at all, Nix will fall back to the following list in impure and unrestricted evaluation mode:
```

No reason to recall it's an environment variable (we have that from context). I suggest to emphasize that being set to empty and not being set is very different.

File: doc/manual/src/command-ref/env-common.md (hunk 2 above)
Comment (zuzuleinen) [reply doc/manual/src/command-ref/env-common.md]: Indeed running  `env NIX_PATH="" nix-build --no-out-link -E 'with import <nixpkgs> { }; go' ` will give me that error


```
error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I)

       at «string»:1:13:

            1| with import <nixpkgs> { }; go
             |             ^
(use '--show-trace' to show detailed location information)
```

File: doc/manual/src/command-ref/env-common.md [hunk 4]
```diff
@@ -44,6 +44,12 @@ Most Nix commands interpret the following environment variables:
     The Nix search path can also be extended using the `-I` option to
     many Nix commands, which takes precedence over `NIX_PATH`.
 
+    If `NIX_PATH` is not set at all, Nix will fall back to the following list in impure and unrestricted evaluation mode:
+
+    1. `$HOME/.nix-defexpr/channels`
+    2. `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```
Comment (jtojnar) [inline doc/manual/src/command-ref/env-common.md:50]:
```suggestion
    2. `nixpkgs=/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

File: doc/manual/src/command-ref/env-common.md [hunk 5]
```diff
@@ -44,6 +44,16 @@ Most Nix commands interpret the following environment variables:
     The Nix search path can also be extended using the `-I` option to
     many Nix commands, which takes precedence over `NIX_PATH`.
 
+    If `NIX_PATH` is not set at all, Nix will fall back to the following list in impure and unrestricted evaluation mode:
+
+      1. `$HOME/.nix-defexpr/channels`
+      2. `nixpkgs=/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
+      3. `/nix/var/nix/profiles/per-user/root/channels`
+
+    If `NIX_PATH` is set to an empty string the above fallback will not happen and building will fail with:
```
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/env-common.md:53]: Thanks for the quick response. Trying to be even more sharp:
```suggestion
    If `NIX_PATH` is set to an empty string, resolving search paths will always fail.
    For example, attempting to use `<nixpkgs>` will produce:
```

File: doc/manual/src/command-ref/env-common.md [hunk 6]
```diff
@@ -44,6 +44,16 @@ Most Nix commands interpret the following environment variables:
...
+
+      `error: file 'nixpkgs' was not found in the Nix search path`
```
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/env-common.md:55]:
```suggestion
Thanks for the quick response. Trying to be even more sharp:
```suggestion
    If `NIX_PATH` is set to an empty string, resolving search paths will always fail.
    For example, attempting to use `<nixpkgs>` will produce:

        error: file 'nixpkgs' was not found in the Nix search path
```

The error message should now render as code.

**Review by fricklerhandwerk — APPROVED** [review]
Great stuff, thanks for your patience @zuzuleinen!

**Comment by amardeep** [comment]
Is this planned to be merged?

File: doc/manual/src/command-ref/env-common.md [hunk 7]
```diff
@@ -44,6 +44,17 @@ Most Nix commands interpret the following environment variables:
     The Nix search path can also be extended using the `-I` option to
     many Nix commands, which takes precedence over `NIX_PATH`.
 
+    If `NIX_PATH` is not set at all, Nix will fall back to the following list in impure and unrestricted evaluation mode:
```
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/env-common.md:47]:
```suggestion
    If `NIX_PATH` is not set at all, Nix will fall back to the following list in [impure](./conf-file.md#conf-pure-eval) and [unrestricted](./conf-file.md#conf-restrict-eval) evaluation mode:
```

**Comment by fricklerhandwerk** [comment]
@amardeep thanks for the reminder. I did not have merge permissions back then, and do not keep track of things I can't do anything about.

**Pull request merged**